src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,118 @@
 package core.basesyntax;
 
-public class ArrayList<T> implements List<T> {
+import java.util.Arrays;
+import java.util.NoSuchElementException;
 
+public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    
+    private T[] elements;
+    private int size;
+    
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+    
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
-
+    
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index);
+        if (index >= elements.length) {
+            grow();
+            add(value, index);
+        } else if (index < size) {
+            if (size + 1 > elements.length) {
+                grow();
+            }
+            T[] firstSubArray = Arrays.copyOfRange(elements, 0, index + 1);
+            T[] secondSubArray = Arrays.copyOfRange(elements, index, elements.length - 1);
+            firstSubArray[firstSubArray.length - 1] = value;
+            T[] resultArray = Arrays.copyOf(firstSubArray,
+                    firstSubArray.length + secondSubArray.length);
+            System.arraycopy(secondSubArray, 0, resultArray,
+                    firstSubArray.length, secondSubArray.length);
+            elements = resultArray;
+            size++;
+        } else {
+            elements[index] = value;
+            size++;
+        }
```

Comment:
let's rewrite this all in 3 lines

1 - you can use your `elements` both as source and destination array in your System.arraycopy
2 - set the value
3 - increase the size